jh update documentation (#198)
AY2223S1-CS2103-F14-3/tp: docs/DeveloperGuide.md

How the `Logic` component works:
1. When `Logic` is called upon to execute a command, it uses the `AddressBookParser` class to parse the user command.
2. This results in a `Command` object (more precisely, an object of one of its subclasses e.g., `AddCommand`) which is executed by the `LogicManager`.

> 3. The command can communicate with the `Model` when it is executed (e.g. to add a person).

Reviewer: ```suggestion
3. The command can communicate with the `Model` when it is executed (e.g. to add an application).
```

* All `XYZCommandParser` classes (e.g., `AddCommandParser`, `DeleteCommandParser`, ...) inherit from the `Parser` interface so that they can be treated similarly where possible e.g, during testing.

### Model component

> **API** : [`Model.java`](https://github.com/AY2223S1-CS2103-F14-3/tp/blob/bb7ac45ed934914d1814ef862edaddf4ccfff320/src/main/java/seedu/application/model/Model.java)

Reviewer 2: Probably shouldn't use the permalink here. This link always directs to a particular commit in the past (the one with hash bb7ac45...) The link that always goes to the latest version can be taken from the address bar

```suggestion
**API** : [`Model.java`](https://github.com/AY2223S1-CS2103-F14-3/tp/blob/master/src/main/java/seedu/application/model/Model.java)
```

* stores the currently 'selected' `Application` objects (e.g., results of a search query) as a separate _filtered_ list which is exposed to outsiders as an unmodifiable `ObservableList<Application>` that can be 'observed' e.g. the UI can be bound to this list so that the UI automatically updates when the data in the list change.
* stores the `Application` objects which have existing `Interview` present in them. It is an `ObservableList<Application>` which is used to create the interview list in the UI.
* stores the `Application` objects which have the `Interview` within one week from the time the user is using the application as another _filtered_ list.
* stores the `Application` objects sorted by different possible orders as a _sorted_ list. It acts as an important list for the user to select the index for `EditCommand`, `DeleteCommand`, etc.
* stores a `UserPref` object that represents the user’s preferences. This is exposed to the outside as a `ReadOnlyUserPref` objects.
* does not depend on any of the other three components (as the `Model` represents data entities of the domain, they should make sense on their own without depending on other components)

> <div markdown="span" class="alert alert-info">:information_source: **Note:** An alternative (arguably, a more OOP) model is given below. It has a `Tag` list in the `ApplicationBook`, which `Application` references. This allows `ApplicationBook` to only require one `Tag` object per unique tag, instead of each `Application` needing their own `Tag` objects. In addition, `Interview` is optional in each `Application`, thus, it is wrapped around `Optional` class to avoid the usage of `null`.<br>

Reviewer: ```suggestion
<div markdown="span" class="alert alert-info">:information_source: **Note:** An alternative (arguably, a more OOP) model is given below. It has a `Tag` list in the `ApplicationBook`, which `Application` references. This allows `ApplicationBook` to only require one `Tag` object per unique tag, instead of each `Application` needing their own `Tag` objects. In addition, `Interview` is optional in each `Application`, thus, it is wrapped around with an `Optional` class to avoid the usage of `null`.<br>
```

    * Pros: Reduced coupling between different classes.
    * Cons: Each time an action is performed on an application or interview (such as archiving, adding or editing), the list of upcoming interviews has to be informed and updated as well. 

### [Proposed] Find Interview Feature

#### Implementation
1. The user enters `find-i technical interview` to filter the existing interviews with keyword `technical interview` in one of the interview fields, e.g. `Round`. The execution prompts the `LogicManager` to call the `ApplicationBookParser#parseCommand(String)` method.

> 2. The ApplicationBookParser then identifies the corresponding `FindInterviewCommandParser` to create. Then, the keywords entered are used to instantiate new `Predicates` which in turns to be used to construct new `FindInterviewCommand`.

Reviewer: ```suggestion
2. The ApplicationBookParser then identifies the corresponding `FindInterviewCommandParser` to create. Then, the keywords entered are used to instantiate new `Predicates` which in turn is to be used to construct new `FindInterviewCommand`.
```

> 2. The ApplicationBookParser then identifies the corresponding `FindInterviewCommandParser` to create. Then, the keywords entered are used to instantiate new `Predicates` which in turns to be used to construct new `FindInterviewCommand`.

Reviewer: ```suggestion
2. The ApplicationBookParser then identifies the corresponding `FindInterviewCommandParser` to create. Then, the keywords entered are used to instantiate new `Predicates` which in turn is to be used to construct new `FindInterviewCommand`.
```
